File: slizaa-hierarchicalgraph/graphdb/org.slizaa.hierarchicalgraph.graphdb.mapping.service/src/main/java/org/slizaa/hierarchicalgraph/graphdb/mapping/service/internal/DefaultMappingService.java
package org.slizaa.hierarchicalgraph.graphdb.mapping.service.internal;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slizaa.core.boltclient.IBoltClient;
import org.slizaa.hierarchicalgraph.core.model.HGRootNode;
import org.slizaa.hierarchicalgraph.core.model.HierarchicalgraphFactory;
import org.slizaa.hierarchicalgraph.core.model.INodeSource;
import org.slizaa.hierarchicalgraph.core.model.impl.ExtendedHGRootNodeImpl;
import org.slizaa.hierarchicalgraph.core.model.spi.INodeComparator;
import org.slizaa.hierarchicalgraph.core.model.spi.IProxyDependencyResolver;
import org.slizaa.hierarchicalgraph.graphdb.mapping.cypher.IBoltClientAware;
import org.slizaa.hierarchicalgraph.graphdb.mapping.service.IMappingParticipator;
import org.slizaa.hierarchicalgraph.graphdb.mapping.service.IMappingService;
import org.slizaa.hierarchicalgraph.graphdb.mapping.service.MappingException;
import org.slizaa.hierarchicalgraph.graphdb.mapping.spi.*;
import org.slizaa.hierarchicalgraph.graphdb.model.GraphDbHierarchicalgraphFactory;
import org.slizaa.hierarchicalgraph.graphdb.model.GraphDbRootNodeSource;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Function;
import java.util.stream.Collectors;

import static com.google.common.base.Preconditions.checkNotNull;
import static org.slizaa.hierarchicalgraph.graphdb.mapping.service.internal.GraphFactoryFunctions.*;

public class DefaultMappingService implements IMappingService {

    private static final Logger log = LoggerFactory.getLogger(DefaultMappingService.class);

    private List<IMappingParticipator> _mappingParticipators = new CopyOnWriteArrayList<>();

    static Function<Long, INodeSource> createNodeSourceFunction = (id) -> {
        INodeSource nodeSource = GraphDbHierarchicalgraphFactory.eINSTANCE.createGraphDbNodeSource();
        nodeSource.setIdentifier(id);
        return nodeSource;
    };

    public void addMappingParticipator(IMappingParticipator mappingParticipator) {
        this._mappingParticipators.add(mappingParticipator);
    }

    public void removeMappingParticipator(IMappingParticipator mappingParticipator) {
        this._mappingParticipators.remove(mappingParticipator);
    }

    @Override
    public HGRootNode convert(IMappingProvider mappingDescriptor, final IBoltClient boltClient) throws MappingException {

        checkNotNull(mappingDescriptor);
        checkNotNull(boltClient);

        try {
            long totalStart = System.currentTimeMillis();
            long stepStart;

            // create the root element
            final HGRootNode rootNode = HierarchicalgraphFactory.eINSTANCE.createHGRootNode();
            rootNode.registerExtension(IBoltClient.class, boltClient);
            GraphDbRootNodeSource rootNodeSource = GraphDbHierarchicalgraphFactory.eINSTANCE.createGraphDbRootNodeSource();
            rootNodeSource.setIdentifier(-1l);
            rootNodeSource.setBoldClient(boltClient);
            rootNode.setNodeSource(rootNodeSource);

            // process root, hierarchy and dependency queries
            stepStart = System.currentTimeMillis();
            IHierarchyDefinitionProvider hierarchyProvider = initializeBoltClientAwareMappingProviderComponent(
                    mappingDescriptor.getHierarchyDefinitionProvider(), boltClient);
            log.info("Initialized hierarchy provider in {}ms", System.currentTimeMillis() - stepStart);

            if (hierarchyProvider != null) {

                stepStart = System.currentTimeMillis();
                List<Long> rootNodes = hierarchyProvider.getToplevelNodeIds();
                createFirstLevelElements(rootNodes.toArray(new Long[0]), rootNode, createNodeSourceFunction);
                log.info("Created {} root nodes in {}ms", rootNodes.size(), System.currentTimeMillis() - stepStart);

                stepStart = System.currentTimeMillis();
                List<Long[]> parentChildNodeIds = hierarchyProvider.getParentChildNodeIds();
                createHierarchy(parentChildNodeIds, rootNode, createNodeSourceFunction);
                log.info("Created hierarchy ({} parent-child pairs) in {}ms", parentChildNodeIds.size(), System.currentTimeMillis() - stepStart);

                stepStart = System.currentTimeMillis();
                removeDanglingNodes(rootNode);
                log.info("Removed dangling nodes in {}ms", System.currentTimeMillis() - stepStart);

                stepStart = System.currentTimeMillis();
                IDependencyDefinitionProvider dependencyProvider = initializeBoltClientAwareMappingProviderComponent(
                        mappingDescriptor.getDependencyDefinitionProvider(), boltClient);
                log.info("Initialized dependency provider in {}ms", System.currentTimeMillis() - stepStart);

                if (dependencyProvider != null) {
                    stepStart = System.currentTimeMillis();
                    var dependencies = dependencyProvider.getDependencies();
                    createDependencies(dependencies, rootNode,
                            (id, type) -> GraphFactoryFunctions.createDependencySource(id, type, null), false);
                    log.info("Created {} dependencies in {}ms", dependencies.size(), System.currentTimeMillis() - stepStart);
                }
            }

            // register default extensions
            rootNode.registerExtension(IProxyDependencyResolver.class, new CustomProxyDependencyResolver());
            rootNode.registerExtension(IMappingProvider.class, mappingDescriptor);
            rootNode.registerExtension(INodeComparator.class, mappingDescriptor.getNodeComparator());
            rootNode.registerExtension(ILabelDefinitionProvider.class, mappingDescriptor.getLabelDefinitionProvider());
            rootNode.registerExtension(INodeMetadataProvider.class, mappingDescriptor.getNodeMetadataProvider());

            for (IMappingParticipator mappingParticipator : this._mappingParticipators) {
                mappingParticipator.postCreate(rootNode, mappingDescriptor, boltClient);
            }

            log.info("Graph conversion completed in {}ms", System.currentTimeMillis() - totalStart);
            return rootNode;
        }
        catch (Exception e) {
            throw new MappingException(e.getMessage(), e);
        }
    }

    private <T> T initializeBoltClientAwareMappingProviderComponent(T component, final IBoltClient boltClient) throws Exception {

        if (component instanceof IBoltClientAware) {
            ((IBoltClientAware) component).initialize(boltClient);
        }

        return component;
    }

    private void removeDanglingNodes(final HGRootNode rootNode) {
        List<Object> nodeKeys2Remove = ((ExtendedHGRootNodeImpl) rootNode).getIdToNodeMap().entrySet().stream()
                .filter((n) -> {
                    try {
                        return !new Long(0).equals(n.getValue().getIdentifier()) && n.getValue().getRootNode() == null;
                    } catch (Exception e) {
                        return true;
                    }
                }).map(n -> n.getKey()).collect(Collectors.toList());
        nodeKeys2Remove.forEach(k -> ((ExtendedHGRootNodeImpl) rootNode).getIdToNodeMap().remove(k));
    }
}

File: slizaa-hierarchicalgraph/graphdb/io.hierograph.hierarchicalgraph.graphdb.mapping.service/src/main/kotlin/io/hierograph/hierarchicalgraph/graphdb/mapping/service/DefaultMappingService.kt
package io.hierograph.hierarchicalgraph.graphdb.mapping.service

import io.hierograph.hierarchicalgraph.core.model.HGNode
import io.hierograph.hierarchicalgraph.core.model.HGRootNode
import io.hierograph.hierarchicalgraph.core.model.HierarchicalGraphFactory
import io.hierograph.hierarchicalgraph.graphdb.mapping.spi.IDependencyDefinition
import io.hierograph.hierarchicalgraph.graphdb.mapping.spi.IMappingProvider
import io.hierograph.hierarchicalgraph.graphdb.mapping.spi.INodeMetadataProvider
import io.hierograph.hierarchicalgraph.graphdb.mapping.spi.ParentChildNode
import io.hierograph.hierarchicalgraph.graphdb.mapping.spi.RootNode
import io.hierograph.hierarchicalgraph.graphdb.model.GraphDbDependencySource
import io.hierograph.hierarchicalgraph.graphdb.model.GraphDbNodeSource
import io.hierograph.hierarchicalgraph.graphdb.model.GraphDbRootNodeSource
import org.slizaa.core.boltclient.IBoltClient

class DefaultMappingService : IMappingService {

    override fun convert(mappingProvider: IMappingProvider, boltClient: IBoltClient): HGRootNode {
        // 1. Create root node
        val rootNodeSource = GraphDbRootNodeSource(identifier = -1L)
        rootNodeSource.boltClient = boltClient
        val rootNode = HierarchicalGraphFactory.createRootNode { rootNodeSource }
        rootNode.registerExtension(IBoltClient::class.java, boltClient)

        // Node lookup map (neo4j id -> HGNode)
        val idToNodeMap = mutableMapOf<Long, HGNode>()

        // 2. Initialize hierarchy provider
        val hierarchyProvider = mappingProvider.hierarchyDefinitionProvider
        if (hierarchyProvider is IBoltClientAware) {
            hierarchyProvider.initialize(boltClient)
        }

        // 3. Build root-level nodes
        val rootNodes = hierarchyProvider.getToplevelNodeIds()
        for (rn in rootNodes) {
            val node = createNodeIfAbsent(rn.id, rootNode, rootNode, idToNodeMap)
            setKindIfNull(node, rn.kind)
        }

        // 4. Build hierarchy
        val parentChildNodes = hierarchyProvider.getParentChildNodeIds()
        for (pcn in parentChildNodes) {
            val parentNode = createNodeIfAbsent(pcn.parentId, rootNode, null, idToNodeMap)
            val childNode = createNodeIfAbsent(pcn.childId, rootNode, parentNode, idToNodeMap)
            setKindIfNull(childNode, pcn.childKind)
        }

        // 5. Remove dangling nodes
        val danglingKeys = idToNodeMap.entries
            .filter { (_, node) -> node.parent == null }
            .map { it.key }
        danglingKeys.forEach { idToNodeMap.remove(it) }

        // 6. Initialize dependency provider
        val dependencyProvider = mappingProvider.dependencyDefinitionProvider
        if (dependencyProvider is IBoltClientAware) {
            dependencyProvider.initialize(boltClient)
        }

        // 7. Build dependencies
        val dependencies = dependencyProvider.getDependencies()
        for (depDef in dependencies) {
            createDependency(depDef, rootNode, idToNodeMap)
        }

        // 8. Register extensions
        rootNode.registerExtension(IMappingProvider::class.java, mappingProvider)
        rootNode.registerExtension(INodeMetadataProvider::class.java, mappingProvider.nodeMetadataProvider)

        // 9. Return
        return rootNode
    }

    private fun createNodeIfAbsent(
        id: Long,
        rootNode: HGRootNode,
        parent: HGNode?,
        idToNodeMap: MutableMap<Long, HGNode>
    ): HGNode {
        val existing = idToNodeMap[id]
        if (existing != null) {
            // Node exists but may need parent set
            if (existing.parent == null && parent != null) {
                HierarchicalGraphFactory.setParent(existing, parent)
            }
            return existing
        }

        // Create new node
        val nodeSource = GraphDbNodeSource(identifier = id)
        val node = HierarchicalGraphFactory.createNode(rootNode, parent ?: rootNode) { nodeSource }
        idToNodeMap[id] = node
        return node
    }

    private fun setKindIfNull(node: HGNode, kind: Any) {
        if (node.kind == null) {
            node.kind = kind
        }
    }

    private fun createDependency(
        depDef: IDependencyDefinition,
        rootNode: HGRootNode,
        idToNodeMap: Map<Long, HGNode>
    ) {
        val fromNode = idToNodeMap[depDef.idStart] ?: return
        val toNode = idToNodeMap[depDef.idTarget] ?: return

        val dep = HierarchicalGraphFactory.createCoreDependency(
            fromNode, toNode, depDef.type,
            { GraphDbDependencySource(identifier = depDef.idRel, type = depDef.type) }
        )
        dep.weight = depDef.weight
        dep.attributesBitmap = depDef.attributesBitmap
    }
}
